### ISSUE TITLE
Incorrect mscorlib selected by msbuild

### ISSUE THREAD

**Issue by citizenmatt** [JetBrains/resharper-unity#988]
When generating the `.csproj` files, Rider 2018.3 will add a hint path to all explicitly referenced system assemblies to point to the current Unity version's Mono install. This usually affects `System`, `System.Xml`, `System.Core`, `System.Runtime.Serialization` and `System.Xml.Linq`.

However, msbuild also adds an explicit reference to `mscorlib`, as well as a secondary reference to `System.Core`. These references have no hint path and are resolved by the target framework version.

[Unity 2017.2](https://github.com/Unity-Technologies/UnityCsReference/commit/9be0e8275aa1aa1ece02d2284bca53c2956699a0#diff-ae1ccef0c7a021d0bcea8ed33e811075R376) to (but not including) [Unity 2018.3.0a9](https://github.com/Unity-Technologies/UnityCsReference/commit/555f53cb7e0b443f6db8fc227f062c73f5b96990#diff-ae1ccef0c7a021d0bcea8ed33e811075L494) (links take a while to fully load) would hard code the target framework version as `v4.5` for Rider, even for a project targeting .NET 2.0/3.5.

This means that we can get into the situation where `System`, `System.Xml`, etc. are pointing at Unity's `2.0-api` Mono folder, and msbuild resolves `mscorlib` against the system installed .NET 4.5. The secondary reference of `System.Core` is auto-unified to the resolved version - 4.5 in the system installed .NET folder.

This can cause compile errors, such as:

```
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/TestRunner.cs(134,13,134,28): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/ConsolePro/Remote/LiteNetLib/NetSerializer.cs(118,9,131,10): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(43,24,43,62): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(67,17,67,46): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(104,17,104,40): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/Trinary Software/Timing.cs(1477,69,1477,80): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
```

(Note that I haven't been able to recreate this error locally. I get unresolved symbols for `ISet<>` in ReSharper and when building. I think you might need to include a reference to something that causes one of the system installed .NET system libraries to be referenced.)

I think this will fix it:

* Set `<NoStdLib>` to `true`. This will tell msbuild to not add the explicit reference to `mscorlib`
* Set `<AddAdditionalExplicitAssemblyReferences>` to `false`. This stops msbuild adding an extra reference to `System.Core` without a hint path, which then gets resolved to the wrong version and auto-unified, overriding the explicit reference in the project
* Add an explicit reference to `mscorlib`, with a hint path that matches `System`

Optionally, we can also do one or more of the following:

* Set `<AssemblySearchPaths>` to `{HintPathFromItem}`. This clears all search paths, and requires that references must have a hint path. Alternatively, we could copy most of these items, but remove `{TargetFrameworkDirectory}` and the `{Registry:…}` entries so that we'll search in normal places, but never resolve against the target framework directory.

    ```
    <!--
        The SearchPaths property is set to find assemblies in the following order:

            (1) Files from current project - indicated by {CandidateAssemblyFiles}
            (2) $(ReferencePath) - the reference path property, which comes from the .USER file.
            (3) The hintpath from the referenced item itself, indicated by {HintPathFromItem}.
            (4) The directory of MSBuild's "target" runtime from GetFrameworkPath.
                The "target" runtime folder is the folder of the runtime that MSBuild is a part of.
            (5) Registered assembly folders, indicated by {Registry:*,*,*}
            (6) Assembly folders from AssemblyFolders.config file (provided by Visual Studio Dev15+).
            (7) Legacy registered assembly folders, indicated by {AssemblyFolders}
            (8) Resolve to the GAC.
            (9) Treat the reference's Include as if it were a real file name.
            (10) Look in the application's output folder (like bin\debug)
        -->
    <AssemblySearchPaths Condition=" '$(AssemblySearchPaths)' == ''">
      {CandidateAssemblyFiles};
      $(ReferencePath);
      {HintPathFromItem};
      {TargetFrameworkDirectory};
      $(AssemblyFoldersConfigFileSearchPath)
      {Registry:$(FrameworkRegistryBase),$(TargetFrameworkVersion),$(AssemblyFoldersSuffix)$(AssemblyFoldersExConditions)};
      {AssemblyFolders};
      {GAC};
      {RawFileName};
      $(OutDir)
    </AssemblySearchPaths>
    ```

* Set the target framework directory to the correct folder. We could probably leave the system assemblies without hint paths and let msbuild find them in the new target framework dir. This would mean revisiting `AvoidGetReferenceAssemblyCalls`. We should probably look at this for pre-Unity 2018.1 too.

**Comment by van800** [comment]
I expect this issue to fire only for Unity prior to 2018.1

**Comment by van800** [comment]
Fix will be public in Rider 2018.3.2.
